docs/proposals/31-inplace-node-update.md

```diff
@@ -0,0 +1,388 @@
+    inPlaceUpdate:
+      maxUnavailable: 1
+      maxSurge: 0
+      onLabel: false # true/false
+    type: InPlaceUpdate # RollingUpdate/InPlaceUpdate/Recreate
+  template: {}
+    spec: {}
+status: {}
+```
+
+`maxUnavailable`: Specifies the maximum number of unavailable machines during the update process.
+
+`maxSurge`: Specifies the maximum number of machines that can be scheduled above the desired number of machines. Set `maxSurge` to zero if you don't want any new nodes to be provisioned during the update. You can set it to a value greater than zero if you don't want unavailable nodes during the update; in that case, first new nodes will be created with the updated configuration and then the old nodes will undergo in-place updates.
+
+`onLabel`: If the update strategy is `InPlaceUpdateOnLabel` in the worker spec, the worker controller will set this field to true.
+
+### Dependency Watchdog
+
+Currently, the [`dependency-watchdog`](https://github.com/gardener/dependency-watchdog) monitors the leases of the nodes in the shoot cluster. If nodes more than a configured threshold are unhealthy, it scales down the `machine-controller-manager`, `kube-controller-manager`, and `cluster-autoscaler` to prevent a meltdown. During an ongoing in-place update, DWD can make use of the labels added by `machine-controller-manager` during the `in-place` update to exclude nodes undergoing updates from its health check.
+
+### Extensions
+
+#### OS extensions
+
+The OS extension that wants to support in-place updates should add an "update-script" to the `.spec.files[]` of the OSC with a pre-defined path that will be executed by the `gardener-node-agent` to perform the update. This script should trigger the OS-provided update tool/utility to update the OS.
+
+### Failures and Recovery Strategy
+
+In-place updates may fail because of several factors ranging from network issues to compatibility problems. This section outlines the failure scenarios, error classifications, and the specific roles `gardener-node-agent` and `machine-controller-manager` play in failure handling and recovery to ensure minimal disruption.
+
+#### Types of Failures During Update
+
+##### Drain Timeout/Failure
+
+- A node is marked as `ready-for-update` only after a successful drain.
+- We use the existing [drain logic](https://github.com/gardener/machine-controller-manager/tree/v0.54.0/pkg/util/provider/drain) provided by the `machine-controller-manager`.
+- If a drain operation times out or fails, the node will be forcefully drained.
+
+##### Kubernetes Update Failures
+
+- **Rollback on Failure:** If a Kubernetes version update fails, `gardener-node-agent` will roll back the `kubelet` to the version specified in the last applied `OperatingSystemConfig`. This ensures that the node remains operational.
+  - **Update Halt on Failure:** `gardener-node-agent` will label the node with an `update-failed` label. This label signals `machine-controller-manager` to stop further updates.
```

Comment:
typically you define an `errorThreshold` during an update. If that gets breached then you stop further updates. `maxUnavailable` can also be used but it will then get overloaded. Its purpose will be two-fold then - one to determine how many number of nodes can be in-place updated concurrently and also to determine when to stop further updates. If you wish to have different threshold value than number of concurrent updates then you cannot just use `maxUnavailable`.

Reply:
Request that `update-failed` have a succinct, error message for the label value that is logable by the MCM and be capable of being put in the `Machine` status.  If this is not ideal, please consider an the use of a separate annotation `node.machine.sapcloud.io/update-message` that can hold the full error message.

Reply:
> Maybe it should be rather maxUnavailable.

It will be maxUnavailble only, MCM will pick a machine to get an update only if it can fulfill `maxUnavailable`, also any machine undergoing an update or already failed an update will be counted in the unhealthy machines.